### PR TITLE
Code cleanup and unit testing overhaul (partial)

### PR DESCRIPTION
**Pull request by nicolassaw**
Biggest updates:
- rewrote the scraper code to use functions that are more modular and easier to follow/read, and also to make it possible to test with tiny unit tests
- created unit tests for the scraper and parser generally
- began creating unit tests for pieces of the scraper, but not all of them

Left to do:
- finish creating unit tests for all of the scraper tests
- rewrite the parser, cleaner, and updater to have more modular functions
- create tiny unit test for each of those individual functions

### CODE REVIEW

**Comment by nicolassaw** [comment]
**Update**

Finished adding the unit tests for the scraper subfunctions.

Left to do:

[X] finish creating unit tests for all of the scraper tests
[ ] rewrite the parser, cleaner, and updater to have more modular functions
[ ] create tiny unit test for each of those individual functions

File: src/scraper/__main__.py [hunk 1]
```diff
@@ -1,35 +1,81 @@
...
+        # These are class scraper-level fields that need to be stored and used from function to function. 
+        self.odyssey_version = None
+        self.base_url = None
+        self.main_soup = None
+        self.main_page_html = None
+        self.search_url = None
+        self.search_page_html = None
+        self.search_soup = None
+        self.hidden_values = None
+        self.judicial_officer_to_ID = None
```
Comment (datlassian) [inline src/scraper/__main__.py:45]: Might be helpful to add type annotations to these values (and the other instance values above).

File: src/scraper/__main__.py [hunk 2]
```diff
@@ -1,35 +1,81 @@
...
+            os.makedirs(self.case_html_path, exist_ok=True)
+            # Makes a list of existing cases on the device
+            self.cached_case_list = [file_name.split(".")[0] for file_name in os.listdir(self.case_html_path)]
+        else: # if test = True
+            self.case_html_path = os.path.join(
+                os.path.dirname(__file__), "..", "..", "resources", 'test_files', 'test_data', self.county, "case_html"
+            )
+            os.makedirs(self.case_html_path, exist_ok=True)
+            # Makes a list of existing cases on the device
+            self.cached_case_list = [file_name.split(".")[0] for file_name in os.listdir(self.case_html_path)]
```
Comment (datlassian) [inline src/scraper/__main__.py:74]: 📓  Just noting here as we discussed in-person, this is probably an ideal place to use Dependency Injection.

File: src/scraper/__main__.py [hunk 3]
```diff
@@ -182,12 +245,133 @@ def scrape(self):
             if option.text
         }
         # if juidicial_officers param is not specified, use all of them
-        if not args.judicial_officers:
-            args.judicial_officers = list(judicial_officer_to_ID.keys())
+        if not judicial_officers:
+            judicial_officers = list(judicial_officer_to_ID.keys())
```
Comment (datlassian) [inline src/scraper/__main__.py:249]: ❓ This is my python inexperience, but does `.keys()` already return a list?

File: src/scraper/__main__.py [hunk 4]
```diff
@@ -182,12 +245,133 @@ def scrape(self):
...
+    def scrape_results_page(self, odyssey_version, base_url, search_url, hidden_values, JO_id, date_string):
+        # POST a request for search results
+        results_page_html = request_page_with_retry(
+            session=self.session,
+            url=search_url
+            if odyssey_version < 2017
+            else urllib.parse.urljoin(base_url, "Hearing/SearchHearings/HearingSearch"),
+            verification_text="Record Count"
+            if odyssey_version < 2017
+            else "Search Results",
```
Comment (datlassian) [inline src/scraper/__main__.py:261]: This block trips me up.. I think this logic could be more clear?

File: src/scraper/__main__.py [hunk 5]
```diff
@@ -196,117 +380,33 @@ def scrape(self):
...
+    def scrape(self):
+        self.base_url, self.odyssey_version, self.notes = self.get_ody_link(self.county)
+        self.main_page_html, self.main_soup = self.scrape_main_page(self.base_url, self.odyssey_version, self.notes)
+        self.search_url, self.search_page_html, self.search_soup = self.scrape_search_page(self.base_url, self.odyssey_version, self.main_page_html, self.main_soup)
+        self.hidden_values = self.scrape_hidden_values(self.odyssey_version, self.main_soup, self.search_soup)
+        if self.case_number:
+            self.scrape_individual_case(self.base_url, self.search_url, self.hidden_values, self.case_number)
+        else: # scrape a list of JOs between a start and end date
+            self.judicial_officers, self.judicial_officer_to_ID = self.scrape_jo_list(self.odyssey_version, self.search_soup, self. judicial_officers)
+            self.START_TIME = time() # initialize variables to time script and build a list of already scraped cases
```
Comment (datlassian) [inline src/scraper/__main__.py:410]: 🔧 May want to rename this `SCRAPE_START_TIME`, or something a little more specific.

Comment (datlassian) [reply src/scraper/__main__.py:410]: It might also just be fine as a local variable in this method, rather than an instance variable

**Review by datlassian — APPROVED** [review]

File: src/scraper/__main__.py (hunk 1 above)
Comment (nicolassaw) [reply src/scraper/__main__.py:45]: Will do!

File: src/scraper/__main__.py (hunk 3 above)
Comment (nicolassaw) [reply src/scraper/__main__.py:249]: It's a view object for a dictionary's keys, but has to be set to a field if you want to call it later. It doesn't change the original field.

File: src/scraper/__main__.py (hunk 4 above)
Comment (nicolassaw) [reply src/scraper/__main__.py:261]: I am making a comment to myself to come back to this and change it, hah.

File: src/scraper/__main__.py (hunk 5 above)
Comment (nicolassaw) [reply src/scraper/__main__.py:410]: I'll take a look to double check if it is used anywhere else or if it's local and then update the field name accordingly.

**Review by nicolassaw — COMMENTED** [review]
Ready for final approval after incorporating or noting Dan's notes.

File: src/scraper/__main__.py (hunk 3 above)
Comment (datlassian) [reply src/scraper/__main__.py:249]: I see! Boy my brain does not think in classes by default 😆

**Review by newswim — APPROVED** [review]

**Pull request merged**